File: graph traversals 2/news_distribution.cpp
#include <cstdio>
#include <vector>

long findSetRoot(long element, std::vector<long> &parent) {
    if (parent[element] == element) {
        return element;
    }
    return parent[element] = findSetRoot(parent[element], parent);
}

int main() {
    long n, m;
    scanf("%ld %ld", &n, &m);

    std::vector<long> parent(n + 1);
    std::vector<long> groupSize(n + 1, 1);
    for (long i = 1; i <= n; i++) {
        parent[i] = i;
    }

    while (m--) {
        long groupCount;
        scanf("%ld", &groupCount);

        if (groupCount <= 0) {
            continue;
        }

        long firstMember;
        scanf("%ld", &firstMember);
        long root = findSetRoot(firstMember, parent);

        while (--groupCount) {
            long member;
            scanf("%ld", &member);
            long memberRoot = findSetRoot(member, parent);

            if (root != memberRoot) {
                // Union of two sets
                parent[memberRoot] = root;
                groupSize[root] += groupSize[memberRoot];
            }
        }
    }

    for (long i = 1; i <= n; i++) {
        long root = findSetRoot(i, parent);
        printf("%ld ", groupSize[root]);
    }
    puts("");

    return 0;
}

File: graph traversals 2/news distribution.py
def find(x, parent):
    if parent[x] != x:
        parent[x] = find(parent[x], parent)
    return parent[x]

n, m = map(int, input().split())
parent = [i for i in range(n + 1)]
group_size = [1] * (n + 1)

for _ in range(m):
    group = list(map(int, input().split()))
    if len(group) <= 1:
        continue

    roots = {find(member, parent) for member in group[1:group[0]+1]}

    main_root = max(roots, key=lambda x: group_size[x])
    for r in roots:
        if main_root != r:
            if group_size[main_root] < group_size[r]:
                main_root, r = r, main_root
            parent[r] = main_root
            group_size[main_root] += group_size[r]

for i in range(1, n + 1):
    root = find(i, parent)
    print(group_size[root], end=" ")
print()
